scheduler/replica_scheduler.go

```diff
@@ -45,13 +45,27 @@ func NewReplicaScheduler(ds *datastore.DataStore) *ReplicaScheduler {
 	return rcScheduler
 }
 
+func isFirstAttachment(rs map[string]*longhorn.Replica) bool {
+	for _, r := range rs {
+		if r.Spec.HealthyAt != "" {
+			return false
+		}
+	}
+	return true
+}
+
 // ScheduleReplica will return (nil, nil) for unschedulable replica
 func (rcs *ReplicaScheduler) ScheduleReplica(replica *longhorn.Replica, replicas map[string]*longhorn.Replica, volume *longhorn.Volume) (*longhorn.Replica, util.MultiError, error) {
 	// only called when replica is starting for the first time
 	if replica.Spec.NodeID != "" {
 		return nil, nil, fmt.Errorf("BUG: Replica %v has been scheduled to node %v", replica.Name, replica.Spec.NodeID)
 	}
 
+	if !isFirstAttachment(replicas) && replica.Spec.HealthyAt == "" && replica.Spec.FailedAt != "" {
+		logrus.WithField("replica", replica.ObjectMeta.Name).Warnf("Replica might not be scheduled.")
+		return nil, nil, nil
+	}
```

Comment:
It seems that somehow Longhorn wrongly re-scheduled a failed replica and ignores the affinity setting. 
But why doesn't the check `replica.Spec.NodeID != ""` prevent that from happening?

Reply:
`replica.Spec.NodeID != ""` does not work because this replica was not scheduled before such as volume degraded. In the issue longhorn/longhorn#4546, there is only one worker node and  only one replica is running and the others are still in scheduling (`replica.Spec.NodeID` is empty but no available node.) Then all replica will be marked as failed after the worker node rebooted and all replica would be scheduled to this worker node because replica marked as failed does not occupy the node.

Reply:
OK, got it. The key point of this PR is to delay the failed but unscheduled replicas before handling the reusable or salvageable replicas, right?

Reply:
Yes.

Reply:
Then you can have a separate function with an easy-to-understand naming for this check.

Reply:
OK, modified. 
Thanks.

Reply:
Make it more simple, just to check the replica `HealtyhAt` and `FailedAt`.